course5/homework09_2/main.cc

```diff
@@ -0,0 +1,116 @@
+/*
+Write a simple single-object implementation of std::unique_ptr, complete with make_unique.
+*/
+#include <iostream>
+#include <string>
+#include <cassert>
+
+
+// operators new/delete or new[]/delete[]
+
+template<typename T>
+struct MoreUniquePtr
+{
+	MoreUniquePtr() = default;
+	MoreUniquePtr(const T& src)
+	{
+		ptr = new T;
+		*ptr = src;
+	}
+	MoreUniquePtr(const MoreUniquePtr&) = delete; // copy
+
+	MoreUniquePtr(MoreUniquePtr&& src) // move
+	{
+		ptr = src.ptr;
+		src.ptr = nullptr;
+	}
+
+	~MoreUniquePtr()
+	{
+		if (ptr)
+			delete ptr;
+	}
+
+	MoreUniquePtr& operator=(const MoreUniquePtr& src) = delete; // copy assignmt
+
+	MoreUniquePtr& operator=(MoreUniquePtr&& src) // move assignmt
+	{
+		if (ptr)
+			delete ptr;
+		ptr = src.ptr;
+		src.ptr = nullptr;
+		return *this;
+	}
+
+	T get()
```

Comment:
This is not what std:unique_ptr is doing.

The semantic you are proposing is the one of a single-item container rather than smart pointer.
You provide functionality to copy data in or out of it. It is almost tempting to add a copy constructor that copies the content between the instances (just like containers do).

Smart pointer simply provides basic memory management while still retaining pointer semantics. It is still a "handle to an object", not a container. User has direct access to "the object" via ```get()```, ```operator*``` and ```operator->```.